### PR TITLE
Fix/caas uniter restart

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Fixes:
1. Uniter should not block the `containerStartChan` in operator;
2. Change to use sharing exec client for uniters;
3. CAAS uniter now can restart itself correctly;

## QA steps

```console
$ juju deploy /tmp/charm-builds/mariadb-k8s/ --debug  --resource mysql_image=mariadb --storage database=100M

## exec into operator to enable wrench
$ mkubectl exec pod/mariadb-k8s-operator-0  -n t1 -it -- bash

## enable wrench
$ mkdir wrench && echo "fatal-error" > wrench/remote-init

## scale application to 2 to create a new unit
$ juju scale-application mariadb-k8s -m t1 2

## from the log we can see uniter tries to restart itself because the wrench returns an error in remote init process
$ juju debug-log -m k1:t1 --color --replay --level WARNING

## remove wrench in the operator pod;
$ rm wrench/remote-init

## now we can see the uniter restarts successfully;

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1878329

### CODE REVIEW

File: worker/uniter/remotestate/watcher.go [hunk 1]
```diff
@@ -204,9 +204,8 @@ func (w *RemoteStateWatcher) CommandCompleted(completed string) {
 	}
 }
 
-func (w *RemoteStateWatcher) setUp(unitTag names.UnitTag) error {
+func (w *RemoteStateWatcher) setUp(unitTag names.UnitTag) (err error) {
```
Comment (wallyworld) [inline worker/uniter/remotestate/watcher.go:207]: don't need this

File: worker/uniter/uniter.go [hunk 2]
```diff
@@ -260,6 +260,22 @@ func newUniter(uniterParams *UniterParams) func() (worker.Worker, error) {
 }
 
 func (u *Uniter) loop(unitTag names.UnitTag) (err error) {
+
+	defer func() {
+		// If this is a CAAS unit, then dead errors are fairly normal ways to exit
+		// the uniter main loop, but the actual agent needs to keep running.
```
Comment (wallyworld) [inline worker/uniter/uniter.go:266]: s/actual agent/parent operator agent

File: worker/uniter/remotestate/watcher.go (hunk 1 above)
Comment (ycliuhw) [reply worker/uniter/remotestate/watcher.go:207]: just removed the `var err error` below, so we need this

**Comment by ycliuhw** [comment]
!!build!!
19:04:31 Error: websocket: close 1006 (abnormal closure): unexpected EOF

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**